{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5978cfb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "import os\n",
    "from os import path\n",
    "from os.path import isfile, join, dirname, isdir, exists\n",
    "\n",
    "\n",
    "import fnmatch\n",
    "import gffpandas.gffpandas as gffpd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def make_dir(*argv):\n",
    "    mydir = path.join(*argv)    \n",
    "    if not path.exists(mydir):        \n",
    "        if len(argv) > 1:\n",
    "            make_dir(*argv[:-1])            \n",
    "        os.mkdir(mydir)\n",
    "    return mydir\n",
    "\n",
    "\n",
    "def make_path(*argv):\n",
    "    mypath = path.join(*argv)\n",
    "    if not path.exists(dirname(mypath)):\n",
    "        make_dir(*argv[:-1])\n",
    "    return mypath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27132e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tool for searching gff table and finding a row given a gene id\n",
    "\n",
    "ecotype = '10015'\n",
    "gene_id = 'T029-R1'\n",
    "\n",
    "annotation = gffpd.read_gff3('gff/%s.gff'%ecotype)\n",
    "\n",
    "df = annotation.df\n",
    "\n",
    "foundGene = False\n",
    "exon_start_l = []\n",
    "exon_end_l = []\n",
    "\n",
    "for ii in range(df.shape[0]):\n",
    "    if df.iloc[ii]['type'] == 'mRNA':\n",
    "        if not foundGene:\n",
    "            if df.iloc[ii]['attributes'].split('|')[1].split(';')[0] == gene_id:\n",
    "                foundGene = True\n",
    "                startGene = df.iloc[ii]['start']\n",
    "                print(ii)\n",
    "        else:\n",
    "            break\n",
    "    elif df.iloc[ii]['type'] == 'exon':\n",
    "        if foundGene:\n",
    "            exon_start_l.append(df.iloc[ii]['start'])\n",
    "            exon_end_l.append(df.iloc[ii]['end'])\n",
    "\n",
    "exon_start_l = [x - startGene for x in exon_start_l]\n",
    "exon_end_l = [x - startGene for x in exon_end_l] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71b4ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ecotype = '10015'\n",
    "annotation = gffpd.read_gff3(make_path('gff', ecotype+'.gff'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dedc121",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(annotation.df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f40bbf05",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = annotation.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7925b466",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49587a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a list of gene id's\n",
    "\n",
    "#list comprehension\n",
    "gene_l = [attr.split('|')[1].split(';')[0] for attr in df['attributes'][df['type']=='mRNA']]\n",
    "\n",
    "#for loop\n",
    "if False:\n",
    "    gene_l = []\n",
    "    for attr in df['attributes'][df['type']=='mRNA']:\n",
    "        gene_l.append(attr.split('|')[1].split(';')[0])\n",
    "        \n",
    "#check if there are no redundancies in gene list\n",
    "len(gene_l) == len(set(gene_l))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0092ae79",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A more efficient way of searching the table. Doesn't find non-contiguous annotations, though.\n",
    "\n",
    "if False:\n",
    "    gene_id = 'T379-R1'\n",
    "\n",
    "    foundGene = False\n",
    "    exon_start_l = []\n",
    "    exon_end_l = []\n",
    "\n",
    "    for ii in range(df.shape[0]):\n",
    "        if df.iloc[ii]['type'] == 'mRNA':\n",
    "            if not foundGene:\n",
    "                if df.iloc[ii]['attributes'].split('|')[1].split(';')[0] == gene_id:\n",
    "                    foundGene = True\n",
    "                    startGene = df.iloc[ii]['start']\n",
    "            else:\n",
    "                break\n",
    "        elif df.iloc[ii]['type'] == 'exon':\n",
    "            if foundGene:\n",
    "                exon_start_l.append(df.iloc[ii]['start'])\n",
    "                exon_end_l.append(df.iloc[ii]['end'])\n",
    "                \n",
    "    exon_start_l = [x - startGene for x in exon_start_l]\n",
    "    exon_end_l = [x - startGene for x in exon_end_l]            \n",
    "\n",
    "    plt.figure(figsize = (10, 3))\n",
    "\n",
    "    for xmin, xmax in zip(exon_start_l, exon_end_l):\n",
    "        plt.hlines(y='', xmin=xmin, xmax=xmax, color='b')                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab9c40e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A more efficient way of searching the table. Doesn't find non-contiguous annotations, though.\n",
    "\n",
    "if True:\n",
    "    gene_id = 'T319-R1'\n",
    "\n",
    "    foundGene = False\n",
    "    exon_start_l = []\n",
    "    exon_end_l = []\n",
    "\n",
    "    for ii in range(df.shape[0]):\n",
    "        if df.iloc[ii]['type'] == 'mRNA':\n",
    "            if not foundGene:\n",
    "                if df.iloc[ii]['attributes'].split('|')[1].split(';')[0] == gene_id:\n",
    "                    foundGene = True\n",
    "                    startGene = df.iloc[ii]['start']\n",
    "                    print(ii)\n",
    "            else:\n",
    "                break\n",
    "        elif df.iloc[ii]['type'] == 'exon':\n",
    "            if foundGene:\n",
    "                exon_start_l.append(df.iloc[ii]['start'])\n",
    "                exon_end_l.append(df.iloc[ii]['end'])\n",
    "                \n",
    "    exon_start_l = [x - startGene for x in exon_start_l]\n",
    "    exon_end_l = [x - startGene for x in exon_end_l]                          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f507bca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[476:].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e3bca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A more efficient way of searching the table. Doesn't find non-contiguous annotations, though.\n",
    "\n",
    "if False:\n",
    "    gene_id = 'T379-R1'\n",
    "\n",
    "    foundGene = False\n",
    "    exon_start_l = []\n",
    "    exon_end_l = []\n",
    "\n",
    "    for ii in range(df.shape[0]):\n",
    "        if df.iloc[ii]['type'] == 'mRNA':\n",
    "            if not foundGene:\n",
    "                if df.iloc[ii]['attributes'].split('|')[1].split(';')[0] == gene_id:\n",
    "                    foundGene = True\n",
    "                    startGene = df.iloc[ii]['start']\n",
    "            else:\n",
    "                break\n",
    "        elif df.iloc[ii]['type'] == 'exon':\n",
    "            if foundGene:\n",
    "                exon_start_l.append(df.iloc[ii]['start'])\n",
    "                exon_end_l.append(df.iloc[ii]['end'])\n",
    "                \n",
    "    exon_start_l = [x - startGene for x in exon_start_l]\n",
    "    exon_end_l = [x - startGene for x in exon_end_l]            \n",
    "\n",
    "    plt.figure(figsize = (10, 3))\n",
    "\n",
    "    for xmin, xmax in zip(exon_start_l, exon_end_l):\n",
    "        plt.hlines(y='', xmin=xmin, xmax=xmax, color='b')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "675a4635",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Searching the table for exons in a gene. Finds non-contiguous annotations.\n",
    "\n",
    "plt.ioff()\n",
    "\n",
    "for gene_id in gene_l:\n",
    "\n",
    "    exon_coord_l = []\n",
    "\n",
    "    for ii in range(df.shape[0]):\n",
    "        if df.iloc[ii]['type'] == 'exon':\n",
    "            if df.iloc[ii]['attributes'].split('|')[1].split(';')[0].startswith(gene_id):\n",
    "                exon_coord_l.append((df.iloc[ii]['start'], df.iloc[ii]['end']))     \n",
    "\n",
    "    geneStart = min([min(coord) for coord in exon_coord_l])\n",
    "    exon_coord_zerostart = [(start-geneStart, end-geneStart) for start,end in exon_coord_l]\n",
    "\n",
    "    fig = plt.figure(figsize = (10, 1))\n",
    "\n",
    "    for xmin, xmax in exon_coord_zerostart:\n",
    "        plt.hlines(y='', xmin=xmin, xmax=xmax, color='b')\n",
    "    fig.tight_layout()        \n",
    "    fig.patch.set_facecolor('xkcd:white') # make transparent background white\n",
    "    fig.savefig(make_path('intron_plots', ecotype, gene_id+'.png'), bbox_inches=\"tight\")\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6927dbf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71be55fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bd8137a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A more efficient way of searching the table. Doesn't find non-contiguous annotations, though.\n",
    "\n",
    "gene_id = 'T378-R1'\n",
    "\n",
    "foundGene = False\n",
    "exon_coord_l = []\n",
    "\n",
    "for ii in range(df.shape[0]):\n",
    "    if df.iloc[ii]['type'] == 'mRNA':\n",
    "        if not foundGene:\n",
    "            if df.iloc[ii]['attributes'].split('|')[1].split(';')[0] == gene_id:\n",
    "                foundGene = True\n",
    "        else:\n",
    "            break\n",
    "    elif df.iloc[ii]['type'] == 'exon':\n",
    "        if foundGene:\n",
    "            exon_coord_l.append((df.iloc[ii]['start'], df.iloc[ii]['end']))   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2751e69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_coord_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c54c8220",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_start_l = []\n",
    "exon_end_l = []\n",
    "\n",
    "gene_exon_start_l = []\n",
    "gene_exon_end_l = []\n",
    "\n",
    "geneStart = 2707\n",
    "geneEnd = 6575\n",
    "\n",
    "for ii in range(df.shape[0]):\n",
    "    if df.iloc[ii]['seq_id'] == 'tig00000074':\n",
    "        if df.iloc[ii]['type'] == 'exon':\n",
    "            exon_start_l.append(df.iloc[ii]['start'])\n",
    "# print(exon_start_l)\n",
    "            exon_end_l.append(df.iloc[ii]['end'])\n",
    "# print(exon_end_l)\n",
    "\n",
    "for x, y in zip(exon_start_l, exon_end_l):\n",
    "    if x >= geneStart and x <= geneEnd: \n",
    "        gene_exon_start_l.append(x)\n",
    "    if y >= geneStart and y <= geneEnd: \n",
    "        gene_exon_end_l.append(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0245085",
   "metadata": {},
   "outputs": [],
   "source": [
    "xs = [1, 100]\n",
    "\n",
    "plt.figure(figsize = (10, 7))\n",
    "\n",
    "plt.vlines(x = gene_exon_start_l, ymin = 0, ymax = max(xs),\n",
    "           colors = 'blue',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = gene_exon_end_l, ymin = 0, ymax = max(xs),\n",
    "           colors = 'red',\n",
    "           label = 'vline_multiple - full height')\n",
    "plt.title(\"ID=10015|G379;Alias=maker-tig00000074-snap-gene-0.6;Name=10015|G379;NLR=TNL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02387f3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_l = []\n",
    "for jj, end in enumerate(gene_exon_end_l):\n",
    "    combined_l.append((gene_exon_start_l[jj], end))\n",
    "# print(combined_l)\n",
    "\n",
    "xs = [1, 100]\n",
    "\n",
    "plt.figure(figsize = (10, 7))\n",
    "\n",
    "plt.vlines(x = combined_l[0], ymin = 0, ymax = max(xs),\n",
    "           colors = 'red',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[1], ymin = 0, ymax = max(xs),\n",
    "           colors = 'orange',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[2], ymin = 0, ymax = max(xs),\n",
    "           colors = 'yellow',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[3], ymin = 0, ymax = max(xs),\n",
    "           colors = 'green',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[4], ymin = 0, ymax = max(xs),\n",
    "           colors = 'blue',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[5], ymin = 0, ymax = max(xs),\n",
    "           colors = 'purple',\n",
    "           label = 'vline_multiple - full height')\n",
    "\n",
    "plt.vlines(x = combined_l[6], ymin = 0, ymax = max(xs),\n",
    "           colors = 'pink',\n",
    "           label = 'vline_multiple - full height')\n",
    "plt.title(\"ID=10015|G379;Alias=maker-tig00000074-snap-gene-0.6;Name=10015|G379;NLR=TNL\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
